### PR TITLE
[Wayland] Fix Showing/Hiding, Placing and Events for Menu Windows

### PR DESCRIPTION
**Pull request by msisov**
This adaptation makes it possible to show/hide and place menu windows
properly in regards to their parent surface. What is more, this patch
fixes event locations sent to the client from wayland.

### CODE REVIEW

File: ui/ozone/platform/wayland/wayland_pointer.h [hunk 1]
```diff
@@ -45,6 +51,10 @@ class WaylandPointer {
                    uint32_t axis,
                    wl_fixed_t value);
 
+  void SetSerial(uint32_t serial);
+
+  WaylandConnection* connection_ = nullptr;
+
```
Comment (tonikitoo) [inline ui/ozone/platform/wayland/wayland_pointer.h]: nit: drop this extra line

File: ui/ozone/platform/wayland/wayland_window.cc [hunk 2]
```diff
@@ -71,10 +114,27 @@ void WaylandWindow::ApplyPendingBounds() {
   connection_->ScheduleFlush();
 }
 
-void WaylandWindow::Show() {}
+void WaylandWindow::Show() {
+  if (xdg_surface_)
```
Comment (tonikitoo) [inline ui/ozone/platform/wayland/wayland_window.cc]: maybe add a comment why only popups need this?

Comment (tonikitoo) [reply ui/ozone/platform/wayland/wayland_window.cc]: actually, why do we create the popup again duplicating lines 83-93 above?

File: ui/ozone/platform/wayland/wayland_window.cc [hunk 3]
```diff
@@ -101,11 +161,19 @@ void WaylandWindow::SetTitle(const base::string16& title) {
...
+  if (old_capture)
+    old_capture->delegate()->OnLostCapture();
+
+  g_current_capture_ = this;
 }
 
 void WaylandWindow::ReleaseCapture() {
-  NOTIMPLEMENTED();
+  if (g_current_capture_ == this)
+    g_current_capture_ = nullptr;
```
Comment (tonikitoo) [inline ui/ozone/platform/wayland/wayland_window.cc:176]: no need to call g_current_capture->delegate()->OnLostCapture(); ?

File: ui/ozone/platform/wayland/wayland_window.cc [hunk 4]
```diff
@@ -149,20 +217,48 @@ PlatformImeController* WaylandWindow::GetPlatformImeController() {
 
 bool WaylandWindow::CanDispatchEvent(const PlatformEvent& native_event) {
   Event* event = static_cast<Event*>(native_event);
-  if (event->IsMouseEvent())
-    return has_pointer_focus_;
+  if (event->IsMouseEvent()) {
+    if (g_current_capture_ == this) {
```
Comment (tonikitoo) [inline ui/ozone/platform/wayland/wayland_window.cc:221]: I would move this 'if' condition into a well named local function. wdyt?

File: ui/ozone/platform/wayland/wayland_window.cc [hunk 5]
```diff
@@ -149,20 +217,48 @@ PlatformImeController* WaylandWindow::GetPlatformImeController() {
...
+  if (event->IsLocatedEvent() && !has_pointer_focus_)
+    ConvertEventToDifferentWindow(event);
+
   DispatchEventFromNativeUiEvent(
       native_event, base::Bind(&PlatformWindowDelegate::DispatchEvent,
                                base::Unretained(delegate_)));
   return POST_DISPATCH_STOP_PROPAGATION;
 }
 
+void WaylandWindow::ConvertEventToDifferentWindow(ui::Event* event) {
```
Comment (tonikitoo) [inline ui/ozone/platform/wayland/wayland_window.cc]: this naming is odd. what does it try to do?

File: ui/ozone/platform/wayland/wayland_window.h [hunk 6]
```diff
@@ -39,6 +39,16 @@ class WaylandWindow : public PlatformWindow, public PlatformEventDispatcher {
   // Set whether this window has keyboard focus and should dispatch key events.
   void set_keyboard_focus(bool focus) { has_keyboard_focus_ = focus; }
 
+  // Set a child of this window. It is very important in case of nested
+  // xdg_popups as long as we must destroy the very last first and only then
+  // its parent.
+  void set_child_window(WaylandWindow* window) { child_window_ = window; }
+
```
Comment (tonikitoo) [inline ui/ozone/platform/wayland/wayland_window.h]: nit: drop extra line

File: ui/ozone/platform/wayland/wayland_window.h [hunk 7]
```diff
@@ -39,6 +39,16 @@ class WaylandWindow : public PlatformWindow, public PlatformEventDispatcher {
...
+  bool has_focus() { return has_pointer_focus_ || has_keyboard_focus_; }
```
Comment (tonikitoo) [inline ui/ozone/platform/wayland/wayland_window.h]: has_keyboard_or_pointer_focus? and move to line 42, right below the other focus methods?

File: ui/ozone/platform/wayland/wayland_window.cc (hunk 5 above)
Comment (msisov) [reply ui/ozone/platform/wayland/wayland_window.cc]: It converts events location to a different window's location.

**Review by tonikitoo — COMMENTED** [review]
a couple of comments, but very close, maksim. good job!

File: ui/ozone/platform/wayland/wayland_window.cc (hunk 4 above)
Comment (msisov) [reply ui/ozone/platform/wayland/wayland_window.cc:221]: if(HasCapture) ?

File: ui/ozone/platform/wayland/wayland_window.cc (hunk 2 above)
Comment (msisov) [reply ui/ozone/platform/wayland/wayland_window.cc]: In Wayland, there is no way to map and unmap windows for showing and hiding them.

One has to destroy xdg_popup and create it again for showing/hiding events

File: ui/ozone/platform/wayland/wayland_window.cc (hunk 3 above)
Comment (msisov) [reply ui/ozone/platform/wayland/wayland_window.cc:176]: Actually, if we have a menu window open and click on a main window (the one in the background), OnLostCapture tells menu_host that we don't have capture any more and that one closes the menu and its submenus

File: ui/ozone/platform/wayland/wayland_window.cc (hunk 2 above)
Comment (tonikitoo) [reply ui/ozone/platform/wayland/wayland_window.cc]: ok. can we move lines 83-94 and 121-128 into a helper?

File: ui/ozone/platform/wayland/wayland_window.cc (hunk 4 above)
Comment (tonikitoo) [reply ui/ozone/platform/wayland/wayland_window.cc:221]: I mean the IF condition in line 222 which is not so easy to read.

File: ui/ozone/platform/wayland/wayland_window.cc (hunk 5 above)
Comment (tonikitoo) [reply ui/ozone/platform/wayland/wayland_window.cc]: TranslateEventToFocusedWindow?

File: ui/ozone/platform/wayland/wayland_window.h (hunk 7 above)
Comment (msisov) [reply ui/ozone/platform/wayland/wayland_window.h]: done

File: ui/ozone/platform/wayland/wayland_window.cc (hunk 5 above)
Comment (msisov) [reply ui/ozone/platform/wayland/wayland_window.cc]: That's the same as in X11WindowOzone and DesktopWindowTreeHostX11. I'll name it the same way - ConvertEventLocationToCurrentWindowLocation

File: ui/ozone/platform/wayland/wayland_window.cc (hunk 4 above)
Comment (msisov) [reply ui/ozone/platform/wayland/wayland_window.cc:221]: done

File: ui/ozone/platform/wayland/wayland_window.cc (hunk 2 above)
Comment (msisov) [reply ui/ozone/platform/wayland/wayland_window.cc]: done

File: ui/ozone/platform/wayland/wayland_pointer.h (hunk 1 above)
Comment (msisov) [reply ui/ozone/platform/wayland/wayland_pointer.h]: done

File: ui/ozone/platform/wayland/wayland_window.cc [hunk 8]
```diff
@@ -15,6 +15,24 @@
...
+
+static WaylandWindow* g_current_capture_ = nullptr;
+
+static const xdg_popup_listener xdg_popup_listener = {
+    &WaylandWindow::PopupDone,
+};
+
+// TODO(msisov, tonikitoo): fix customization according to screen resolution
+// once we are able to get global coordinates of wayland windows.
+gfx::Rect CustomizeBoundsToScreenResolution(const gfx::Rect& child_bounds,
```
Comment (tonikitoo) [inline ui/ozone/platform/wayland/wayland_window.cc]: TranslateBoundsToScreenCoordinates?

File: ui/ozone/platform/wayland/wayland_window.cc [hunk 9]
```diff
@@ -149,20 +217,47 @@ PlatformImeController* WaylandWindow::GetPlatformImeController() {
...
+  if (event->IsLocatedEvent() && !has_pointer_focus_)
+    ConvertEventLocationToCurrentWindowLocation(event);
+
   DispatchEventFromNativeUiEvent(
       native_event, base::Bind(&PlatformWindowDelegate::DispatchEvent,
                                base::Unretained(delegate_)));
   return POST_DISPATCH_STOP_PROPAGATION;
 }
 
+void WaylandWindow::ConvertEventLocationToCurrentWindowLocation(
```
Comment (tonikitoo) [inline ui/ozone/platform/wayland/wayland_window.cc:246]: Add a TODO abou the duplication with the existing X11 routines

**Review by tonikitoo — APPROVED** [review]
lgtm w/ comments

File: ui/ozone/platform/wayland/wayland_window.cc (hunk 9 above)
Comment (msisov) [reply ui/ozone/platform/wayland/wayland_window.cc:246]: It's already been added in .h file :)

File: ui/ozone/platform/wayland/wayland_window.cc (hunk 8 above)
Comment (msisov) [reply ui/ozone/platform/wayland/wayland_window.cc]: done

**Pull request merged**